File: app/src/main/java/com/popalay/cardme/utils/DialogFactory.java
package com.popalay.cardme.utils;

import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;

import com.popalay.cardme.R;

public final class DialogFactory {

    public static Dialog createSimpleOkErrorDialog(Context context, String message) {
        final AlertDialog.Builder alertDialog = new AlertDialog.Builder(context, R.style.Cardme_Dialog)
                .setTitle(R.string.error)
                .setMessage(message)
                .setPositiveButton(R.string.action_ok, null);
        return alertDialog.create();
    }

    public static Dialog createSimpleInfoDialog(Context context, String message,
            @Nullable DialogInterface.OnDismissListener onDismiss) {
        final AlertDialog.Builder alertDialog = new AlertDialog.Builder(context, R.style.Cardme_Dialog)
                .setMessage(message)
                .setOnDismissListener(onDismiss)
                .setPositiveButton(R.string.action_ok, null);
        return alertDialog.create();
    }

    public static Dialog createSimpleOkCancelDialog(Context context, String message,
            @Nullable DialogInterface.OnClickListener positiveButtonClickListener,
            @Nullable DialogInterface.OnDismissListener onDismissListener) {
        final AlertDialog.Builder alertDialog = new AlertDialog.Builder(context, R.style.Cardme_Dialog)
                .setMessage(message)
                .setPositiveButton(R.string.action_ok, positiveButtonClickListener)
                .setOnDismissListener(onDismissListener)
                .setNegativeButton(R.string.cancel, null);
        return alertDialog.create();
    }

    public static Dialog createCustomButtonsDialog(Context context, String message,
            String positiveButton, String negativeButton,
            @Nullable DialogInterface.OnClickListener positiveButtonClickListener,
            @Nullable DialogInterface.OnDismissListener onDismissListener) {
        final AlertDialog.Builder alertDialog = new AlertDialog.Builder(context, R.style.Cardme_Dialog)
                .setMessage(message)
                .setPositiveButton(positiveButton, positiveButtonClickListener)
                .setOnDismissListener(onDismissListener)
                .setNegativeButton(negativeButton, null);
        return alertDialog.create();
    }

    public static ProgressDialog createProgressDialog(Context context, String message) {
        final ProgressDialog progressDialog = new ProgressDialog(context, R.style.Cardme_Dialog_Progress);
        progressDialog.setMessage(message);
        progressDialog.setCancelable(false);
        return progressDialog;
    }
}

File: app/src/main/kotlin/com/popalay/cardme/utils/DialogFactory.kt
package com.popalay.cardme.utils

import android.app.Dialog
import android.app.ProgressDialog
import android.content.Context
import android.support.annotation.StringRes
import android.support.v7.app.AlertDialog
import com.popalay.cardme.R

object DialogFactory {

    fun createSimpleOkErrorDialog(context: Context, message: String): Dialog {
        return AlertDialog.Builder(context, R.style.Cardme_Dialog)
                .setTitle(R.string.error)
                .setMessage(message)
                .setPositiveButton(R.string.action_ok, null)
                .create()
    }

    fun createSimpleInfoDialog(context: Context, message: String,
                               onDismiss: (() -> Unit)? = null): Dialog {
        return AlertDialog.Builder(context, R.style.Cardme_Dialog)
                .setMessage(message)
                .setOnDismissListener { onDismiss?.invoke() }
                .setPositiveButton(R.string.action_ok, null)
                .create()
    }

    fun createSimpleOkCancelDialog(context: Context, message: String? = null,
                                   inPositive: (() -> Unit)? = null,
                                   onDismiss: (() -> Unit)? = null): Dialog {
        return createCustomButtonsDialog(context, message,
                context.getString(R.string.action_ok),
                context.getString(R.string.action_cancel),
                inPositive, onDismiss)
    }

    fun createCustomButtonsDialog(context: Context, @StringRes message: Int = 0,
                                  @StringRes positiveButton: Int = 0,
                                  @StringRes negativeButton: Int = 0,
                                  inPositive: (() -> Unit)? = null,
                                  onDismiss: (() -> Unit)? = null): Dialog {
        return createCustomButtonsDialog(context,
                if (message > 0) context.getString(message) else null,
                if (positiveButton > 0) context.getString(positiveButton) else null,
                if (negativeButton > 0) context.getString(negativeButton) else null,
                inPositive, onDismiss)
    }

    fun createCustomButtonsDialog(context: Context, message: String? = null,
                                  positiveButton: String? = null,
                                  negativeButton: String? = null,
                                  inPositive: (() -> Unit)? = null,
                                  onDismiss: (() -> Unit)? = null): Dialog {
        return AlertDialog.Builder(context, R.style.Cardme_Dialog)
                .setMessage(message)
                .setPositiveButton(positiveButton) { _, _ -> inPositive?.invoke() }
                .setOnDismissListener { onDismiss?.invoke() }
                .setNegativeButton(negativeButton, null)
                .create()
    }

    fun createProgressDialog(context: Context, @StringRes message: Int = 0): ProgressDialog {
        return createProgressDialog(context, if (message > 0) context.getString(message) else null)
    }

    fun createProgressDialog(context: Context, message: String? = null): ProgressDialog {
        return ProgressDialog(context, R.style.Cardme_Dialog_Progress).apply {
            setMessage(message)
            setCancelable(false)
        }
    }
}
